### PR TITLE
Reduce the size of the static library through changes to options

### PR DESCRIPTION
**Pull request by nilayvaish**

### CODE REVIEW

File: core/options.c [hunk 1]
```diff
@@ -485,28 +552,71 @@ parse_liststring_t(liststring_t *var, void *value) {
     (*var)[MAX_LIST_OPTION_LENGTH-1] = '\0';
 }
 
+static void
+parse_fn(enum option_type_t type, void *ptr1, void *ptr2)
```
Comment (derekbruening) [inline core/options.c]: Not sure I understand the name: it doesn't return a function nor parse a function?  Maybe "invoke_parser_for_type()" or "parse_by_type()"?

File: core/options.c [hunk 2]
```diff
@@ -485,28 +552,71 @@ parse_liststring_t(liststring_t *var, void *value) {
...
+      case OPTION_TYPE_pathstring_t:
+          parse_pathstring_t(ptr1, ptr2);
+          break;
+      case OPTION_TYPE_liststring_t:
+          parse_liststring_t(ptr1, ptr2);
+          break;
+    }
+}
+
+static NOINLINE void
```
Comment (derekbruening) [inline core/options.c]: Best to explain the NOINLINE in a comment

File: core/options.c [hunk 3]
```diff
@@ -485,28 +552,71 @@ parse_liststring_t(liststring_t *var, void *value) {
...
+set_bool_opt(const char *opt, const char *command_line_option,
+             bool *value_true, bool *value_false, void **value)
+{
+    if (!strcmp(opt+1, command_line_option)) {
```
Comment (derekbruening) [inline core/options.c]: style: explicit boolean conversion: == 0
See https://github.com/DynamoRIO/dynamorio/wiki/Code-Style

Below too

File: core/options.c [hunk 4]
```diff
@@ -485,28 +552,71 @@ parse_liststring_t(liststring_t *var, void *value) {
...
+    }
+}
+
+static NOINLINE void
+set_nonbool_opt(const char *opt, const char *command_line_option,
+                const char *optstr, const char **pos,
+                char *wordbuffer, int max_option_length, void **value)
+{
+    if (!strcmp(opt+1, command_line_option)) {
+      *value = getword(optstr, pos, wordbuffer, max_option_length);
```
Comment (derekbruening) [inline core/options.c]: style: indentation

File: core/options.c [hunk 5]
```diff
@@ -485,28 +552,71 @@ parse_liststring_t(liststring_t *var, void *value) {
...
+          /* FIXME: check argument */
+    }
+}
+
+static NOINLINE void
+modifier_fn(int index, options_t *options, bool for_this_process)
+{
+  int j = 0;
```
Comment (derekbruening) [inline core/options.c]: indentation

File: core/options.c [hunk 6]
```diff
@@ -485,28 +552,71 @@ parse_liststring_t(liststring_t *var, void *value) {
...
+                char *wordbuffer, int max_option_length, void **value)
+{
+    if (!strcmp(opt+1, command_line_option)) {
+      *value = getword(optstr, pos, wordbuffer, max_option_length);
+          /* FIXME: check argument */
+    }
+}
+
+static NOINLINE void
+modifier_fn(int index, options_t *options, bool for_this_process)
```
Comment (derekbruening) [inline core/options.c]: How about a function name like "run_option_command()"?

File: core/options.c [hunk 7]
```diff
@@ -485,28 +552,71 @@ parse_liststring_t(liststring_t *var, void *value) {
...
+{
+  int j = 0;
+#  define OPTION_COMMAND(type, name, default_value, command_line_option, \
```
Comment (derekbruening) [inline core/options.c]: #define, and #include below (unless there's some ifdef around this)

File: core/options.c [hunk 8]
```diff
@@ -533,20 +642,26 @@ set_dynamo_options_common(options_t *options, const char *optstr, bool for_this_
...
+            int i = 0;
+            for (i = 0; i < num_options; ++i) {
+                if (option_traits[i].type == OPTION_TYPE_bool) {
+                    set_bool_opt(opt, option_traits[i].name, &value_true,
+                                 &value_false, &value);
+                } else {
+                    set_nonbool_opt(opt, option_traits[i].name, optstr, &pos,
+                                    wordbuffer, sizeof(wordbuffer), &value);
+                }
+                if (value) {
```
Comment (derekbruening) [inline core/options.c]: != NULL

File: core/options.c [hunk 9]
```diff
@@ -533,20 +642,26 @@ set_dynamo_options_common(options_t *options, const char *optstr, bool for_this_
...
+                    void* optptr = (void*)(options) + option_traits[i].offset;
+                    parse_fn(option_traits[i].type, optptr, value);
+                    modifier_fn(i, options, for_this_process);
+                    break;
+                }
+            }
+            if (value) continue;
```
Comment (derekbruening) [inline core/options.c]: != NULL and continue on new line

File: core/options.c [hunk 10]
```diff
@@ -618,57 +730,220 @@ check_param_bounds(uint *val, uint min, uint max, const char *name)
     return ret;
 }
 
-/* Print an option string from an option struct.
- * Xref case 7939, in DEBUG builds the ? : creates an unshared implicit local
- * that led to huge stack usage for update_dynamic_options() so we use
- * methods instead of macros for those.  While we could make these inline
- * (which DEBUG will ignore and in release will result in the same code as
- * MACRO version) these are way off the hot path and this way cuts the
- * size of the release build dll by ~7kb.
```
Comment (derekbruening) [inline core/options.c]: This comment seems very relevant and the gist of it should be kept: there's a reason we're doing all these functions.

File: core/options.c [hunk 11]
```diff
@@ -618,57 +730,220 @@ check_param_bounds(uint *val, uint min, uint max, const char *name)
...
+    snprintf(optionbuff, MAX_OPTION_LENGTH, "-%s '%s' ", option, (*(const pathstring_t*)val_ptr));
+}
+static void
+PRINT_STRING_liststring_t(char *optionbuff, const void *val_ptr, const char *option)
+{
+    snprintf(optionbuff, MAX_OPTION_LENGTH, "-%s '%s' ", option, (*(const liststring_t*)val_ptr));
+}
+
+static void
+print_fn(enum option_type_t type, char *optionbuff, const void *val_ptr,
```
Comment (derekbruening) [inline core/options.c]: Again I would vote for "print_option_type()" or sthg as a name

File: core/options.c [hunk 12]
```diff
@@ -618,57 +730,220 @@ check_param_bounds(uint *val, uint min, uint max, const char *name)
...
+         const char *option)
+{
+    switch(type)
+    {
+      case OPTION_TYPE_bool:
```
Comment (derekbruening) [inline core/options.c]: Ditto on switch style

File: core/options.c [hunk 13]
```diff
@@ -618,57 +730,220 @@ check_param_bounds(uint *val, uint min, uint max, const char *name)
...
+          PRINT_STRING_pathstring_t(optionbuff, val_ptr, option);
+          break;
+      case OPTION_TYPE_liststring_t:
+          PRINT_STRING_liststring_t(optionbuff, val_ptr, option);
+          break;
+    }
+}
+
+static int
+DIFF_bool(const void *ptr1, const void *ptr2)
```
Comment (derekbruening) [inline core/options.c]: Optional: drop to lower case (was upper b/c macro -- lower matches func style name) (ditto for others above)

File: core/options.c [hunk 14]
```diff
@@ -618,57 +730,220 @@ check_param_bounds(uint *val, uint min, uint max, const char *name)
...
+static int
+DIFF_liststring_t(const void *ptr1, const void *ptr2)
+{
+    const char* val1 = (const char*)ptr1;
+    const char* val2 = (const char*)ptr2;
+    return strcmp(val1, val2);
+}
+
+static int
+diff_fn(enum option_type_t type, const void *ptr1, const void *ptr2)
```
Comment (derekbruening) [inline core/options.c]: Ditto on name and switch style

File: core/options.c [hunk 15]
```diff
@@ -618,57 +730,220 @@ check_param_bounds(uint *val, uint min, uint max, const char *name)
...
+static void
+COPY_liststring_t(void *ptr1, const void *ptr2)
+{
+    char* val1 = (char*)ptr1;
+    const char* val2 = (const char*)ptr2;
+    strncpy(val1, val2, sizeof(liststring_t));
+}
+
+static void
+copy_fn(enum option_type_t type, void *ptr1, const void *ptr2)
```
Comment (derekbruening) [inline core/options.c]: Ditto on name and switch style

File: core/options.c [hunk 16]
```diff
@@ -677,21 +952,18 @@ get_dynamo_options_string(options_t *options, char *opstr, int len, bool minimal
...
-            PRINT_STRING_##type(optionbuff, options->name,              \
-                                command_line_option);                   \
-            NULL_TERMINATE_BUFFER(optionbuff);                          \
-            strncat(opstr, optionbuff, len-strlen(opstr)-1);            \
-        }                                                               \
+    if (!minimal) {
+        int i;
+        for (i = 0; i < num_options; ++i) {
+            if (option_traits[i].name[0] != ' ') { /* not synthethic */
+                const void* val_ptr = (void*)(options) + option_traits[i].offset;
```
Comment (derekbruening) [inline core/options.c]: style: s/void* v/void *v/

File: core/options.c [hunk 17]
```diff
@@ -708,20 +980,17 @@ get_pcache_dynamo_options_string(options_t *options, char *opstr, int len,
...
-            PRINT_STRING_##type(optionbuff, options->name,              \
-                                command_line_option);                   \
-            NULL_TERMINATE_BUFFER(optionbuff);                          \
-            strncat(opstr, optionbuff, len-strlen(opstr)-1);            \
-        }                                                               \
+    int i;
+    for (i = 0; i < num_options; ++i) {
+        if (option_traits[i].affects_pcache >= pcache_effect &&
+            option_traits[i].name[0] != ' ') { /* not synthethic */
+            const void* val_ptr = (void*)(options) + option_traits[i].offset;
```
Comment (derekbruening) [inline core/options.c]: style: s/void* v/void *v/

File: core/options.c [hunk 18]
```diff
@@ -765,38 +1031,36 @@ update_dynamic_options(options_t *options, options_t *new_options)
...
-                        "Updating dynamic options : Ignoring static option change "\
-                        "(%.*s changed to %.*s)\n",                     \
-                        MAX_LOG_LENGTH/2-80, optionbuff,                \
-                        MAX_LOG_LENGTH/2-80, new_optionbuff);           \
-                }                                                       \
-            });                                                         \
-        }                                                               \
+    int i;
+    for (i = 0; i < num_options; ++i) {
+        void* val1 = (void*)(options) + option_traits[i].offset;
```
Comment (derekbruening) [inline core/options.c]: More * style here and below

File: core/options.c [hunk 19]
```diff
@@ -765,38 +1031,36 @@ update_dynamic_options(options_t *options, options_t *new_options)
...
+        const void* val2 = (void*)(new_options) + option_traits[i].offset;
+
+        if (OPTION_MOD_DYNAMIC == option_traits[i].modifier) {
+            if (diff_fn(option_traits[i].type, val1, val2)) {
+                copy_fn(option_traits[i].type, val1, val2);
+                updated++;
+            }
+        } else {
+          DOLOG(2, LOG_TOP, {
+              if (option_fns[i].diff_fn(options->name, new_options->name)) {
```
Comment (derekbruening) [inline core/options.c]: ?  What is option_fns?  I think this is a mistake?

File: core/options.c [hunk 20]
```diff
@@ -533,20 +642,26 @@ set_dynamo_options_common(options_t *options, const char *optstr, bool for_this_
...
+            for (i = 0; i < num_options; ++i) {
+                if (option_traits[i].type == OPTION_TYPE_bool) {
+                    set_bool_opt(opt, option_traits[i].name, &value_true,
+                                 &value_false, &value);
+                } else {
+                    set_nonbool_opt(opt, option_traits[i].name, optstr, &pos,
+                                    wordbuffer, sizeof(wordbuffer), &value);
+                }
+                if (value) {
+                    void* optptr = (void*)(options) + option_traits[i].offset;
```
Comment (derekbruening) [inline core/options.c]: .offset is in bytes but this is adding in void* units -- gets the wrong answer, right?

File: core/options.c (hunk 16 above)
Comment (derekbruening) [inline core/options.c]: pointer arith looks off again

File: core/options.c (hunk 17 above)
Comment (derekbruening) [inline core/options.c]: pointer arith again

File: core/options.c [hunk 21]
```diff
@@ -732,22 +1001,19 @@ get_pcache_dynamo_options_string(options_t *options, char *opstr, int len,
...
-                       description, flag, pcache) {                     \
-        if (command_line_option[0] != ' ' && /* not synthethic */       \
-            (op_pcache_t) OPTION_AFFECTS_PCACHE_##name == pcache_effect && \
-            DIFF_##type(options->name,default_options.name)) {          \
-            return true;                                                \
-        }                                                               \
+    int i;
+    for (i = 0; i < num_options; ++i) {
+        if (option_traits[i].affects_pcache == pcache_effect) {
+            const void* val1 = (void*)(options) + option_traits[i].offset;
```
Comment (derekbruening) [inline core/options.c]: style and pointer arith: here and below

**Comment by derekbruening** [comment]
I had started reviewing and had most of these comments before this was closed so I went ahead and finished reviewing: what's the reason for closing?

**Comment by derekbruening** [comment]
Also if you could update your git author first+last name

**Comment by nilayvaish** [comment]
Hi Derek

I had forked the main dynamorio repository and I was making all the changes on the master branch.  I think this is not the suggested way of making changes when using Git.  So I stripped some of the changes (including those in this pull request) from the forked repository, merged with the master and added back my changes on a different branch.  That stripping resulted in this pull request getting closed automatically.

I will make the edits you suggested and create a pull request again.

Thanks
Nilay

**Pull request closed without merging**